Zend/zend_partial.c

```diff
@@ -0,0 +1,589 @@
+
+    if ((parent = zend_partial_parent(&call->This))) {
+        function  = &parent->func;
+        prototype = &parent->prototype;
+
+        partial->argc = parent->argc + ZEND_CALL_NUM_ARGS(call);
+        partial->argv = ecalloc(partial->argc, sizeof(zval));
+
+        memcpy(partial->argv, parent->argv, parent->argc * sizeof(zval));
+
+        uint32_t arg = 0,
+                 aend = parent->argc;
+        zval *param = ZEND_CALL_ARG(call, 1),
+             *pend = param + ZEND_CALL_NUM_ARGS(call);
+
+        while (arg < aend && param < pend) {
+            if (Z_TYPE(partial->argv[arg]) == IS_UNDEF) {
+                if (!Z_ISUNDEF_P(param)) {
+                    ZVAL_COPY_VALUE(&partial->argv[arg], param);
+                }
+                param++;
+            }
+            arg++;
+        }
+
+        if ((ZEND_CALL_INFO(call) & ZEND_CALL_HAS_EXTRA_NAMED_PARAMS)) {
+            if (parent->named) {
+                partial->named = zend_array_dup(parent->named);
+
+                zend_hash_merge(partial->named, call->extra_named_params, zval_copy_ctor, true);
+            } else {
+                partial->named = call->extra_named_params;
+                GC_ADDREF(call->extra_named_params);
+            }
+        }
+    } else {
+        function = call->func;
+
+        partial->argc = ZEND_CALL_NUM_ARGS(call);
+        partial->argv = ecalloc(partial->argc, sizeof(zval));
+
+        memcpy(partial->argv, ZEND_CALL_ARG(call, 1), partial->argc * sizeof(zval));
+        
+        if ((ZEND_CALL_INFO(call) & ZEND_CALL_HAS_EXTRA_NAMED_PARAMS)) {
+            partial->named = call->extra_named_params;
+            GC_ADDREF(call->extra_named_params);
+        }
+    }
+
+    memcpy(&partial->func, function, (function->type == ZEND_INTERNAL_FUNCTION) ?
+              sizeof(zend_internal_function) : sizeof(zend_op_array));
+
+    partial->func.common.fn_flags &= ~ZEND_ACC_CLOSURE;
+    partial->func.common.fn_flags |= ZEND_ACC_IMMUTABLE;
+    
+    if (partial->func.common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE) {
+        partial->func.common.fn_flags |= ZEND_ACC_TRAMPOLINE_PERMANENT;
+    }
+
+    if (info & ZEND_APPLY_FACTORY) {
+        ZVAL_PTR(&partial->This, Z_OBJCE(call->This));
+
+        GC_ADD_FLAGS(Z_OBJ(call->This), IS_OBJ_DESTRUCTOR_CALLED);
+        OBJ_RELEASE(Z_OBJ(call->This));
+        
+        ZEND_ASSERT(partial->func.common.fn_flags & ZEND_ACC_CTOR);
+    } else {
+        ZVAL_COPY(&partial->This, &call->This);
+
+        if (info & ZEND_APPLY_PASS) {
+            partial->func.common.fn_flags |= ZEND_ACC_CTOR;
+        } else {
+            partial->func.common.fn_flags &= ~ZEND_ACC_CTOR;
+        }
+    }
+
+    if (partial->func.type == ZEND_USER_FUNCTION) {
+        zend_op_array *ops = (zend_op_array*) &partial->func;
+        
+        if (ops->static_variables) {
+			ops->static_variables = zend_array_dup(ops->static_variables);
```

Comment:
This doesn't look right to me. Partials should be sharing static vars with their original function. It would be good to add some tests for that.

Reply:
Was awaiting some input here ... will get too it ...